### PR TITLE
Update to the latest version of Matryoshka.

### PR DESCRIPTION
**Pull request by sellout**
There have been a ton of upstream changes (notably directly-recursive
support), so this is not a small update.

### CODE REVIEW

**Comment by sellout** [comment]
There are a ton of small changes, the lazy ones are like `Recursive` -> `RecursiveT`, but that sort of change is also necessary for the way we fake mutual recursion. The less lazy ones are like `[T[_[_]]: Recursive, F[_]]` -> `[T, F[_]](implicit T: Recursive.Aux[T, F])`, which we’ll be pushing through the code incrementally.

Also, all the conversion between `Free[F, A]` and `T[CoEnv[A, F, ?]]` is gone now (except for one case). So no more `toCoEnv` or `freeCataM` or whatever.

File: mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala [hunk 1]
```diff
@@ -1770,68 +1846,68 @@ object WorkflowBuilder {
...
-              }) :: nt)))
-        case SpliceBuilderF(src, structure) =>
-          NonTerminal("SpliceBuilder" :: nodeType, None,
-            render(src) :: structure.map(RC.render))
-        case ArraySpliceBuilderF(src, structure) =>
-          NonTerminal("ArraySpliceBuilder" :: nodeType, None,
-            render(src) :: structure.map(RC.render))
-      }
-    }
+  // implicit def WorkflowBuilderRenderTree[F[_]: Coalesce: Functor]
```
Comment (drostron) [inline mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala]: Any reason to keep this around as a comment?

**Comment by drostron** [comment]
A spiffy update.

File: frontend/src/main/scala/quasar/sql/ast.scala [hunk 2]
```diff
@@ -169,8 +169,8 @@ object Sql {
     }
   }
 
-  implicit def toExprAlgebraOps[A](a: Algebra[Sql, A]): AlgebraOps[Sql, A] =
-    toAlgebraOps[Sql, A](a)
+  // implicit def toExprAlgebraOps[A](a: Algebra[Sql, A]): AlgebraOps[Sql, A] =
```
Comment (sellout) [inline frontend/src/main/scala/quasar/sql/ast.scala]: Yep. This was a workaround for a bug in an old version of Matryoshka, and yay, it’s fixed.

File: frontend/src/main/scala/quasar/std/structural.scala [hunk 3]
```diff
@@ -352,10 +359,13 @@ trait StructuralLib extends Library {
     Func.Input2(Top, Top),
     noSimplification,
     // new Func.Simplifier {
```
Comment (sellout) [inline frontend/src/main/scala/quasar/std/structural.scala]: Good question … I think this is a valid simplification if the first argument (currently ignored) is actually `ShiftMapKeys` of the same input map. So,
```scala
case InvokeUnapply(
  Sized(Embed(InvokeUnapply(ShiftMapKeys, Sized(Embed(map1))))),
  Sized(Embed(InvokeUnapply(ShiftMap, Sized(Embed(map2))))))
    if map1 ≟ map2 =>
  map1.some
```
But I don’t think that’s something I want to play with in this PR.

File: mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala (hunk 1 above)
Comment (sellout) [inline mongodb/src/main/scala/quasar/physical/mongodb/workflowbuilder.scala]: Nope. It should be gone.

File: sparkcore/src/main/scala/quasar/physical/sparkcore/fs/queryfile.scala [hunk 4]
```diff
@@ -77,11 +78,11 @@ object queryfile {
       for {
         qs <- QueryFile.convertToQScriptRead[Fix, FileSystemErrT[PhaseResultT[Free[S, ?],?],?], QScriptRead[Fix, ?]](lc)(lp)
           .map(simplifyRead[Fix, QScriptRead[Fix, ?], QScriptShiftRead[Fix, ?], SparkQScript].apply(_))
-        optQS = qs.transAna(
+        optQS = qs.transHylo(
```
Comment (sellout) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/queryfile.scala]: Ok, I _might_ have solved this. If so, it’ll be in this PR.

Comment (sellout) [inline sparkcore/src/main/scala/quasar/physical/sparkcore/fs/queryfile.scala]: Ok, so I pushed a fix … we’ll see if it breaks any of the connectors …

**Comment by drostron** [comment]
Nice to see `finalizeQScript`. 👍🏾, will merge upon Travis approval.

**Comment by sellout** [comment]
Ok, reverted that over-optimistic fix 😢

**Comment by drostron** [comment]
I'll keep hope for that idea's future. Stale comment pruning was also reverted.

**Comment by drostron** [comment]
Needs a changelog.

**Comment by sellout** [comment]
*sigh* Ok, changelog’d

**Pull request merged**